### PR TITLE
[Pauli6] Support for PauliWord and PauliSentence in qml.dot

### PR DESCRIPTION
**Pull request by Qottmann**
Work in progress, some concepts like how best to treat identities need some fine tuning.

see also [Pauli1](https://github.com/PennyLaneAI/pennylane/pull/4989), [Pauli2](https://github.com/PennyLaneAI/pennylane/pull/5001), [Pauli3](https://github.com/PennyLaneAI/pennylane/pull/5003), [Pauli4](https://github.com/PennyLaneAI/pennylane/pull/5017), on top of which this PR builds; as well as [Pauli5](https://github.com/PennyLaneAI/pennylane/pull/5018)

### CODE REVIEW

**Comment by Qottmann** [comment]
[sc-53283]

**Comment by Qottmann** [comment]
Ideally, we should be able to just have dot do
```python
if pauli:
if all(isinstance(pauli, (PauliWord, PauliSentence)) for pauli in ops):
    res = _dot_pure_paulis(coeffs, ops)

else:
    res = _dot_with_ops_and_paulis(coeffs, ops)

return res if pauli else res.operation()
```

But some edge cases and quirks prohibit that at the moment.

**Comment by Qottmann** [comment]
Actually the above suggestion works. 
It is a trade off though: Removing the performance improvements for making special cases when all `coeffs` have the same magnitude vs. adding the performance improvement of treating everything as pauli sentences internally.

It is a controversial change, feel free to discuss.

**Comment by Qottmann** [comment]
I think there is still a need for more refinement on how `pw.operation()` and `ps.operation()` is handling empty pauli words and sentences that represent identities. Currently we raise an `ValueError`. This PR just treats the case of a PauliSentence explicitly containing (just) an empty PauliWord to make it work. But there is more to be done.

**Comment by Qottmann** [comment]
The failing jax test seems unrelated to this PR and comes from current `master`

File: pennylane/ops/functions/dot.py
Comment (albi3ro) [inline pennylane/ops/functions/dot.py:118]: I'm a bit hesitant about choosing the wire `0` for the identity.  We could potentially allow `Identity` to exist on "no wires" like with do with `GlobalPhase`. That might eliminate this problem of not knowing what wires to use here.

Comment (Qottmann) [reply pennylane/ops/functions/dot.py:118]: That would be amazing!

Comment (Qottmann) [reply pennylane/ops/functions/dot.py:118]: How to best proceed from here: Leave this PR open and blocked until that feature request is merged or continue with that as "fix-later"?

Comment (mudit2812) [reply pennylane/ops/functions/dot.py:118]: I don't have the full context to know this, but do we want people to use `PauliWord({})`? If not, we don't really need to address this

Comment (Qottmann) [reply pennylane/ops/functions/dot.py:118]: This is user facing. E.g. even when you initialize `PauliWord({0:"I"})` it will automatically strip all identity entries and you are left with `PauliWord({})` upon initialization. I think it makes sense that the empty PauliWord is 1 and the empty PauliSentence is 0.

File: pennylane/ops/functions/dot.py
Comment (astralcai) [inline pennylane/ops/functions/dot.py:109]: This `if` statement seems redundant.

Comment (Qottmann) [reply pennylane/ops/functions/dot.py:109]: I think you are right. I recall testing some timings for checking instances versus doing the conversion and seeing a significant difference, but I cant reproduce that anymore. Will remove

```
from pennylane.pauli.pauli_arithmetic import PauliWord, PauliSentence
qml.operation.enable_new_opmath()
X, Y, Z = qml.PauliX, qml.PauliY, qml.PauliZ
ops = [0.5 * (X(i) @ Y(i) @ Z(i)) for i in range(20)]

%timeit any(isinstance(op, (PauliWord, PauliSentence)) for op in ops)
%timeit [op.operation() if isinstance(op, (PauliWord, PauliSentence)) else op for op in ops]
```
```
1.15 µs ± 5.42 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.07 µs ± 4.11 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

File: pennylane/pauli/pauli_arithmetic.py
Comment (mudit2812) [inline pennylane/pauli/pauli_arithmetic.py:804]: I don't like this property very much. I'm assuming this is added so that we can use `qml.pauli.pauli_sentence` with any op given to `qml.dot()`. I think it's more sensible to update `pauli_sentence` directly to handle `PauliWord` and `PauliSentence` using instance checks.

Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py:804]: The motivation here is to do the check `if op.pauli_rep:` though technically this is not essential to this PR actually.

Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py]: Removed this property since its not actually part of this feature. I think it makes sense to add anyway at another point though.

File: pennylane/pauli/pauli_arithmetic.py
Comment (mudit2812) [inline pennylane/pauli/pauli_arithmetic.py:169]: small white space change that I think needs to be resolved?

Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py]: Yep thanks!

Comment (Qottmann) [reply pennylane/pauli/pauli_arithmetic.py]: Yep! Good catch, also fixed some other rendering issues that snuck in

**Review by mudit2812 — APPROVED** [review]
Looks good to me!

**Review by astralcai — APPROVED** [review]
Looks good!

**Comment by albi3ro** [comment]
Screenshot 2024-02-09 at 10 24 32 AM

**Pull request merged**
